docs/posts/duckdb-for-rag/index.qmd

```diff
@@ -0,0 +1,512 @@
+6. Setup the data to insert into the `ibis` table.
+7. Create an Ibis memtable from the data.
+8. If the table does not exist, create it and insert the data.
+9. If the table exists, insert the data.
+10. Drop any memtable views that were created.
+11. List the tables, showing the `ibis` table.
+
+```{python}
+#| code-fold: true
+#| echo: false
+# this is here to avoid building all the docs in this post...
+_ = shutil.rmtree(repo_name, ignore_errors=True)
+```
+
+Let's inspect the table:
+
+```{python}
+t = rag_con.table(repo_name)  # <1>
+t
+```
+
+1. Get the `ibis` table from the DuckDB database.
+
+Now we have a DuckDB database with the `ibis` table containing all the file
+paths and contents of the Quarto markdown files in the `ibis` repository. You
+can adjust this code to work with any repository and any file type(s)! For
+instance, we may also want to ingest all the `.py` files in the Ibis repsoitory.
+
+### Embedding text to vectors
+
+To search for similar text in traditional RAG fashion, we need to **embed** the
+text into vectors. We can use [OpenAI's embedding
+models](https://platform.openai.com/docs/guides/embeddings/embedding-models) to
+perform this step.
+
+Note that the maximum length of input **tokens** is 8,191. A word typically
+consists of 1-3 tokens. To estimate the number of tokens of a string in Python,
+we can divide the length of that string by 4. This isn't perfect, but will work
+for now. Let's augmented our data with the estimated number of tokens and sort
+from largest to smallest:
+
+```{python}
+@ibis.udf.scalar.python  # <1>
+def estimate_tokens(text: str) -> int:  # <1>
+    return len(text) // 4  # <1>
+```
+
+1. Create an Ibis UDF using the Python function.
+
+```{python}
+t = (
+    t.mutate(tokens_estimate=estimate_tokens(t.contents))  # <1>
+    .order_by(ibis._["tokens_estimate"].desc()) # <2>
+    .relocate("filepath", "tokens_estimate") # <3>
+)
+t
+```
+
+1. Augment the table with the estimated number of tokens.
+2. Sort the table by the estimated number of tokens in descending order.
+3. Relocate the `filepath` and `tokens_estimate` columns to the front of the table.
+
+The longest text is certainly over our embedding model's token limit, and the
+second longest might be as well. In practice, we'd want to chunk our text into
+smaller pieces. This is its own challenge, but for the purposes of this
+demonstration we'll just ignore text that is too long.
+
+Let's define our embedding functions:
+
+```{python}
+def _embed(text: str) -> list[float]:  # <1>
+    """Text to fixed-length array embedding."""  # <1>
+    try:  # <2>
+        model = "text-embedding-3-small"  # <3>
+        text = text.replace("\n", " ")  # <4>
+        client = OpenAI()  # <5>
+        return (  # <6>
+            client.embeddings.create(input=[text], model=model).data[0].embedding
+        )  # <6>
+    except Exception as e:  # <7>
+        print(e)  # <7>
+        return None  # <7>
```

Comment:
```suggestion
    model = "text-embedding-3-small"  # <3>
    text = text.replace("\n", " ")  # <4>
    client = OpenAI()  # <5>
    try:  # <2>
        return (  # <6>
            client.embeddings.create(input=[text], model=model).data[0].embedding
        )  # <6>
    except Exception as e:  # <7>
        print(e)  # <7>
        return None  # <7>
```

Seems like the intent of catching the exception is to deal with failed API calls. Moving the client setup and call argument setup out of the except block might help clarify that.